iso8601/parse.go

```diff
@@ -0,0 +1,106 @@
+package iso8601
+
+import (
+	"encoding/binary"
+	"time"
+	"unsafe"
+)
+
+// Parse parses an ISO8601 string, e.g. "2021-03-25T21:36:12Z".
+func Parse(input string) (time.Time, error) {
+	switch len(input) {
+	case 20:
+		b := unsafeStringToBytes(input)
+		_ = b[19] // compiler hint
+
+		t1 := binary.LittleEndian.Uint64(b)
+		t2 := binary.LittleEndian.Uint64(b[8:])
+		t3 := binary.LittleEndian.Uint32(b[16:])
+
+		// Check for valid "YYYY-MM-DDTHH:MM:SSZ" separators by masking input
+		// with "    -  -  T  :  :  Z". If separators are all valid, replace
+		// them with a '0' (0x30) byte and check all bytes are now numeric.
+		// If it's not input we can handle on the fast path, pass it off to
+		// time.Parse().
+		if (t1&sepMask1u64) != sepMask1u64 ||
+			(t2&sepMask2u64) != sepMask2u64 ||
+			(t3&sepMask3u32) != sepMask3u32 ||
+			nonNumeric64(t1 & ^sepMask1u64 | sepZero1u64) != 0 ||
+			nonNumeric64(t2 & ^sepMask2u64 | sepZero2u64) != 0 ||
+			nonNumeric32(t3 & ^sepMask3u32 | sepZero3u32) != 0 {
+
+			goto fallback
+		}
+
+		// TODO: there's probably a faster way to extract the integers, e.g. see
+		//  https://kholdstare.github.io/technical/2020/05/26/faster-integer-parsing.html
```

Comment:
This article is amazing 😍  Thanks for sharing!

Reply:
I didn't end up using the divide & conquer technique, but it could likely be used in the JSON parser to speed up `parse(u)Int`.